Problem_Set_3/ps3.py

```diff
@@ -0,0 +1,516 @@
+        print("Ran out of letters." + "\n" + "Total score for this hand:", total_score)
+        print("----------")
+    return total_score
+
+    
+    # BEGIN PSEUDOCODE <-- Remove this comment when you implement this function
+    # Keep track of the total score
+    
+    # As long as there are still letters left in the hand:
+    
+        # Display the hand
+        
+        # Ask user for input
+        
+        # If the input is two exclamation points:
+        
+            # End the game (break out of the loop)
+
+            
+        # Otherwise (the input is not two exclamation points):
+
+            # If the word is valid:
+
+                # Tell the user how many points the word earned,
+                # and the updated total score
+
+            # Otherwise (the word is not valid):
+                # Reject invalid word (print a message)
+                
+            # update the user's hand by removing the letters of their inputted word
+    
+
+    # Game is over (user entered '!!' or ran out of letters),
+    # so tell user the total score
+
+    # Return the total score as result of function
+
+# play_hand({'h': 1, 'e': 2, 'l': 2, 'y': 1, 'm': 2, '*': 1}, load_words())
+
+#
+# Problem #6: Playing a game
+# 
+
+
+#
+# procedure you will use to substitute a letter in a hand
+#
+
+def substitute_hand(hand, letter):
+    """ 
+    Allow the user to replace all copies of one letter in the hand (chosen by user)
+    with a new letter chosen from the VOWELS and CONSONANTS at random. The new letter
+    should be different from user's choice, and should not be any of the letters
+    already in the hand.
+
+    If user provide a letter not in the hand, the hand should be the same.
+
+    Has no side effects: does not mutate hand.
+
+    For example:
+        substitute_hand({'h':1, 'e':1, 'l':2, 'o':1}, 'l')
+    might return:
+        {'h':1, 'e':1, 'o':1, 'x':2} -> if the new letter is 'x'
+    The new letter should not be 'h', 'e', 'l', or 'o' since those letters were
+    already in the hand.
+    
+    hand: dictionary (string -> int)
+    letter: string
+    returns: dictionary (string -> int)
+    """
+    ## Misinterpreted version of question, where performing as many separate random replacements as the value of
+    ## letter being replaced. Used deepcopy here as workaround.
+
+    # amount = hand.get(letter, 0)
+    # # print(hand)
+    # new_hand = copy.deepcopy(hand)
+    # # print(amount, new_hand)
+    # if letter in hand.keys():
+    #     new_hand.pop(letter)
+    #     # print(hand)
+    #     while amount > 0:
+    #         substitute_letter = random.choice(VOWELS + CONSONANTS)
+    #         if substitute_letter not in hand.keys() and substitute_letter != letter:
+    #             new_hand[substitute_letter] = new_hand.get(substitute_letter, 0) + 1
+    #             amount -= 1
+    #             # print(substitute_letter, new_hand)
+    # return new_hand
+
+    ## Solved properly using for loop. 

```

Comment:
what makes you saay for loop is the right way? :P

Reply:
because the deepcopy sounds like cheating. It says dont change the thing. So i make a copy of the thing and then change that instead, so I'm technically not changing the thing.😂